docs/compilers/terrapin.md

```diff
@@ -0,0 +1,125 @@
+.NET and Terrapin Process
+===
+The Terrapin system is concerned with validating the provenance of NuPkgs in the .NET ecosystem. Specifically it needs to verify that DLLs in a NuPkg can be rebuilt using publicly available artifacts. 
+
+This document describes the workflow that will be used by Terrapin for validating the provenance of .NET binaries, responsibilities of each team, as well as the tools the .NET team will deliver to enable the service to be successful.
+
+## .NET Tools
+The .NET Team will deliver a set of tools to the Terrapin team to provide the necessary validation. The idea in providing a series of tools here vs. a single all in one tool is to help better separate out the different responsibilities involved here and letting the appropriate team handle them:
+
+1. Providing the list of artifacts necessary for validation to succeed (owned by .NET)
+1. Acquiring the artifacts necessary for validation (owned by Terrapin)
+1. Performing validation and making a determination of provenance (owned by .NET)
+
+It's possible in the future we will end up providing a "all in one" experience tool but for the short term we'll be focusing on keeping these responsibilities and tools separate. 
+
+### dotnet-roslyn-manifest-generator
+This is global tool which takes in a DLL / PDB file combination and will generate an artifacts manifest file which lists the set of artifacts necessary to perform a validation. This includes the version of `dotnet-roslyn-validator` which needs to be used to perform the validation. 
+
+This tool is meant to work with binaries produced by earlier versions of the .NET compiler. Every time the contents of the PDB provenance metadata change a new version of this tool will be produced that can account for the metadata. The expectation is that Terrapin will always have the latest version of this tool installed on their service.
+
+### dotnet-roslyn-validator
```

Comment:
Should this rather be two tools? `dotnet-csharp-validator` and `dotnet-visual-basic-validator`, so that we can avoid dependency on both C# and VB from a single tool? The manifest would list the name of the language and the version of the compiler. Using these two inputs the end-to-end tool would chose the right validator package. This would also be applicable to `dotnet-fsharp-validator`, etc.

Reply:
In the long run I'm not sure it will be as simple as single validator. Consider that to validate a single file package we will need to use both a language specific validator as well as an IL validator. This doesn't really argue against having language specific validators but do think the story is more complicated than a single one per PE file.

Reply:
Yes, there might be additional validators. But there will always be a language specific one to do the compilation. The name and version of a package that contains that validator needs to be inferable from the common options stored in the PDB, so that the uber-validator can find it. Other options in the PDB could indicate additional validators which the uber-validator needs to locate and execute.